dbms/src/Storages/Transaction/FastAddPeerCache.cpp

```diff
@@ -0,0 +1,164 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/FailPoint.h>
+#include <Encryption/PosixRandomAccessFile.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/SharedContexts/Disagg.h>
+#include <Storages/Page/V3/CheckpointFile/CPManifestFileReader.h>
+#include <Storages/Page/V3/PageDirectory.h>
+#include <Storages/Page/V3/Universal/RaftDataReader.h>
+#include <Storages/Page/V3/Universal/UniversalPageIdFormatImpl.h>
+#include <Storages/Page/V3/Universal/UniversalPageStorage.h>
+#include <Storages/Page/V3/Universal/UniversalWriteBatchImpl.h>
+#include <Storages/S3/CheckpointManifestS3Set.h>
+#include <Storages/S3/S3Common.h>
+#include <Storages/S3/S3Filename.h>
+#include <Storages/S3/S3GCManager.h>
+#include <Storages/S3/S3RandomAccessFile.h>
+#include <Storages/Transaction/CheckpointInfo.h>
+#include <Storages/Transaction/FastAddPeerAsyncTasksImpl.h>
+#include <Storages/Transaction/KVStore.h>
+#include <Storages/Transaction/Region.h>
+#include <Storages/Transaction/TMTContext.h>
+#include <fmt/core.h>
+
+namespace DB
+{
+UInt64 SegmentEndKeyCache::getSegmentIdContainingKey(const DM::RowKeyValue & key)
+{
+    std::unique_lock lock(mu);
+    if (!is_ready)
+        cv.wait(lock, [&] { return is_ready; });
```

Comment:
Also `cv.wait` is not necessary. Please use `lock` instead.

Reply:
Fixed.